File: iosApp/iosApp/Views/Clubs/MembersTab.swift
import SwiftUI
import Shared
import DesignSystem

struct MembersTab: View {
    let members: [Shared.MemberListItemInfo]
    var participants: [Shared.SessionParticipantInfo] = []
    var currentUserId: String = ""
    var userRole: Shared.Role? = nil
    var onChangeRole: (String) -> Void = { _ in }
    var onRemoveMember: (String) -> Void = { _ in }
    var onTransferOwnership: (String) -> Void = { _ in }
    var onInviteMember: () -> Void = {}

    private var isAdminOrAbove: Bool { userRole == .owner || userRole == .admin }
    private var isOwner: Bool { userRole == .owner }
    private var readingByMemberId: [String: Bool] {
        Dictionary(uniqueKeysWithValues: participants.map { ($0.memberId, $0.isReading) })
    }
    private var showEmptyState: Bool { members.count <= 1 && isAdminOrAbove }

    var body: some View {
        if members.isEmpty {
            NoTabData(text: String(localized: "empty_no_members"))
        } else {
            VStack(alignment: .leading, spacing: 0) {
                HStack {
                    Text("\(members.count) members")
                        .kluvsStyle(KluvsTheme.typography.title.small, feature: true)
                        .foregroundColor(KluvsTheme.colors.contentMuted)
                    Spacer()
                    // The EmptyState below already carries its own "Invite Members" action, so
                    // this would just be a redundant second invite button.
                    if isAdminOrAbove && !showEmptyState {
                        OutlinedButton(text: "+ Invite", action: onInviteMember)
                    }
                }
                .padding(.bottom, 12)

                if showEmptyState {
                    // At most one member — nothing to scroll, so the row (if any) plus the
                    // EmptyState below just fill the tab directly. A ScrollView also proposes
                    // unbounded height to its content, so it can't be used here or the
                    // EmptyState gets stuck at its own small intrinsic size instead.
                    memberRows
                    EmptyState(
                        heading: "Just you, for now.",
                        body: "Invite a few people and this club starts to feel like one."
                    ) {
                        SecondaryButton(text: "Invite Members", action: onInviteMember)
                    }
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                } else {
                    ScrollView {
                        memberRows
                    }
                }
            }
            .padding(16)
        }
    }

    private var memberRows: some View {
        ForEach(Array(members.enumerated()), id: \.element.memberId) { index, member in
            let isSelf = member.userId == currentUserId
            MemberListItem(
                member: member,
                isSelf: isSelf,
                isReading: readingByMemberId[member.memberId],
                showAdminActions: isAdminOrAbove && (!isSelf || isOwner) && member.role != .owner,
                showRemove: isOwner && !isSelf && member.role != .owner,
                showTransferOwnership: isOwner && !isSelf && member.role != .owner,
                onChangeRole: { onChangeRole(member.memberId) },
                onRemove: { onRemoveMember(member.memberId) },
                onTransferOwnership: { onTransferOwnership(member.memberId) }
            )

            if index < members.count - 1 {
                Divider()
            }
        }
    }
}

// MARK: - Member List Item

private struct MemberListItem: View {
    let member: Shared.MemberListItemInfo
    var isSelf: Bool = false
    var isReading: Bool? = nil
    var showAdminActions: Bool = false
    var showRemove: Bool = false
    var showTransferOwnership: Bool = false
    var onChangeRole: () -> Void = {}
    var onRemove: () -> Void = {}
    var onTransferOwnership: () -> Void = {}

    var body: some View {
        HStack(alignment: .center, spacing: 14) {
            Avatar(name: member.name, avatarUrl: member.avatarUrl, size: 40, memberId: member.memberId, isOwn: isSelf)

            // Identity column carries as much weight as the right rail's two tiers instead of
            // looking sparse next to it — matches Android's MemberListItem.
            VStack(alignment: .leading, spacing: 2) {
                HStack(spacing: 8) {
                    Text(member.name)
                        .kluvsStyle(KluvsTheme.typography.body.large)
                        .foregroundColor(KluvsTheme.colors.content)
                    if isSelf {
                        Text("YOU")
                            .kluvsStyle(KluvsTheme.typography.eyebrow)
                            .foregroundColor(KluvsTheme.colors.accent)
                    }
                }
                Text("@\(member.handle)")
                    .kluvsStyle(KluvsTheme.typography.body.medium)
                    .foregroundColor(KluvsTheme.colors.contentMuted)
            }
            .padding(.vertical, 8)

            Spacer()

            VStack(alignment: .trailing, spacing: 0) {
                HStack(spacing: 8) {
                    if let isReading {
                        // 16pt icon inside 8pt padding = 32pt total box, matching Android's
                        // Modifier.size(32.dp).padding(8.dp) (fixed outer box, smaller visible icon).
                        Icon(
                            type: .reading,
                            contentDescription: isReading ? "Reading" : "Skipping",
                            tint: isReading ? KluvsTheme.colors.accent : KluvsTheme.colors.contentMuted
                        )
                        .frame(width: 16, height: 16)
                        .padding(8)
                    }
                    if showAdminActions || showRemove || showTransferOwnership {
                        ActionMenu(items: {
                            var items: [ActionMenuItem] = []
                            if showAdminActions {
                                items.append(ActionMenuItem(label: "Change Role", action: onChangeRole))
                            }
                            if showTransferOwnership {
                                items.append(ActionMenuItem(label: "Make Owner", action: onTransferOwnership))
                            }
                            if showRemove {
                                items.append(ActionMenuItem(label: "Remove", action: onRemove, isDestructive: true))
                            }
                            return items
                        }())
                    }
                }
                if member.role != .member {
                    RoleEyebrow(role: member.role)
                } else {
                    Spacer().frame(height: 16)
                }
            }
        }
        .padding(.vertical, 12)
    }
}

#Preview {
    MembersTab(members: [])
}

File: composeApp/src/androidMain/kotlin/com/ivangarzab/kluvs/ui/clubs/MembersTab.kt
package com.ivangarzab.kluvs.ui.clubs

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.PreviewLightDark
import androidx.compose.ui.unit.dp
import com.ivangarzab.kluvs.R
import com.ivangarzab.kluvs.clubs.presentation.MemberListItemInfo
import com.ivangarzab.kluvs.clubs.presentation.SessionParticipantInfo
import com.ivangarzab.kluvs.model.Role
import com.ivangarzab.kluvs.designsystem.theme.KluvsTheme
import com.ivangarzab.kluvs.designsystem.theme.feature
import com.ivangarzab.kluvs.designsystem.components.avatars.Avatar
import com.ivangarzab.kluvs.designsystem.components.buttons.OutlinedButton
import com.ivangarzab.kluvs.designsystem.components.buttons.PrimaryButton
import com.ivangarzab.kluvs.designsystem.components.icons.Icon
import com.ivangarzab.kluvs.designsystem.components.icons.IconType
import com.ivangarzab.kluvs.designsystem.components.menus.ActionMenu
import com.ivangarzab.kluvs.designsystem.components.menus.ActionMenuItem
import com.ivangarzab.kluvs.designsystem.components.NoTabData
import com.ivangarzab.kluvs.ui.components.RoleEyebrow

@Composable
fun MembersTab(
    modifier: Modifier = Modifier,
    members: List<MemberListItemInfo>,
    participants: List<SessionParticipantInfo> = emptyList(),
    currentUserId: String? = null,
    userRole: Role? = null,
    onChangeRole: (memberId: String) -> Unit = {},
    onRemoveMember: (memberId: String) -> Unit = {},
    onInviteMember: () -> Unit = {},
) {
    // Session participation lookup for the reading/skipping indicator
    val readingByMemberId = participants.associate { it.memberId to it.isReading }
    if (members.isEmpty()) {
        NoTabData(
            modifier = modifier,
            text = R.string.no_members_in_club
        )
        return
    }

    val isAdminOrAbove = userRole == Role.OWNER || userRole == Role.ADMIN
    val isOwner = userRole == Role.OWNER

    Column(modifier = modifier) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = stringResource(R.string.x_members, members.size),
                color = KluvsTheme.colors.contentMuted,
                style = KluvsTheme.typography.title.small.feature()
            )
            if (isAdminOrAbove) {
                OutlinedButton(
                    text = stringResource(R.string.invite),
                    onClick = onInviteMember
                )
            }
        }

        Spacer(Modifier.height(12.dp))

        LazyColumn {
            itemsIndexed(members) { index, member ->
                //TODO: Consider creating an ext function: Member.isMe(Member or Member.id)
                val isSelf = member.userId != null && member.userId == currentUserId
                MemberListItem(
                    memberId = member.memberId,
                    name = member.name,
                    handle = member.handle,
                    avatarUrl = member.avatarUrl,
                    role = member.role,
                    isSelf = isSelf,
                    isReading = readingByMemberId[member.memberId],
                    showAdminActions = isAdminOrAbove && (!isSelf || isOwner),
                    showRemove = isOwner && !isSelf && member.role != Role.OWNER,
                    onChangeRole = { onChangeRole(member.memberId) },
                    onRemove = { onRemoveMember(member.memberId) }
                )
                if (index < members.size - 1) {
                    MemberDivider()
                }
            }
        }

        if (members.size <= 1 && isAdminOrAbove) {
            Spacer(Modifier.height(20.dp))
            Column(
                modifier = Modifier.fillMaxWidth(),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Text(
                    text = stringResource(R.string.invite_others_cta),
                    style = KluvsTheme.typography.headline.small.feature(),
                    color = KluvsTheme.colors.contentMuted,
                    textAlign = TextAlign.Center
                )
                Spacer(Modifier.height(16.dp))
                PrimaryButton(
                    text = stringResource(R.string.invite_members),
                    onClick = onInviteMember,
                )
            }
        }
    }
}

@Composable
private fun MemberDivider() {
    HorizontalDivider(color = KluvsTheme.colors.cardAlt)
}

@PreviewLightDark
@Composable
fun Preview_MembersTab() = KluvsTheme {
    MembersTab(
        modifier = Modifier
            .background(color = KluvsTheme.colors.card)
            .fillMaxSize(),
        members = listOf(
            MemberListItemInfo("0", "Iván Garza Bermea", "ivangarzab", "", role = Role.OWNER, userId = "u0"),
            MemberListItemInfo("1", "Monica Michelle Morales", "monica", "", role = Role.ADMIN, userId = "u1"),
            MemberListItemInfo("2", "Marco \"Chitho\" Rivera", "chitho23", "", role = Role.MEMBER, userId = "u2"),
            MemberListItemInfo("3", "Anacleto \"Keto\" Longoria", "keto92", "", role = Role.MEMBER, userId = "u3"),
            MemberListItemInfo("4", "Joel Oscar Julian Salinas", "josalinas", "", role = Role.MEMBER, userId = "u4"),
            MemberListItemInfo("5", "Ginseng Joaquin Guzman", "gino1", "", role = Role.MEMBER, userId = "u5"),
        ),
        currentUserId = "u0",
        userRole = Role.OWNER
    )
}
